### PR TITLE
Initial integration of hip f8 types

### PR DESCRIPTION
**Pull request by cgmillette**
- Replace previous f8 implementation with hip f8.
- Explicitly differentiates f8/bf8 from fnuz types.
- Adds numeric_limits and some type traits for new f8 types
- Re-organize type traits on non-native datatypes to follow their implementations [e.g., xfloat32]
- Re-enable fp8 tests on wmma
- Fixes first GEMM test run due to warmup - reinitializes resources from args
- Add types to tests and skips tests for new types on archs that don't support them

### CODE REVIEW

File: library/include/rocwmma/internal/float8.hpp [hunk 1]
```diff
@@ -24,948 +24,1047 @@
...
     {
-        return a;
-    }
+    public:
+        static constexpr bool is_signed         = true;
+        static constexpr bool is_integer        = false;
+        static constexpr bool is_specialized    = true;
+        static constexpr bool is_exact          = false;
+        static constexpr bool has_infinity      = true;
+        static constexpr bool has_quiet_NaN     = false;
```
Comment (CongMa13) [inline library/include/rocwmma/internal/float8.hpp]: e5m2 has nan.

File: library/include/rocwmma/internal/float8.hpp [hunk 2]
```diff
@@ -24,948 +24,1047 @@
...
+    public:
+        static constexpr bool is_specialized    = true;
+        static constexpr bool is_signed         = true;
+        static constexpr bool is_integer        = false;
+        static constexpr bool is_exact          = false;
+        static constexpr bool has_infinity      = false;
+        static constexpr bool has_quiet_NaN     = true;
+        static constexpr bool has_signaling_NaN = false;
+        static constexpr auto has_denorm        = true;
+        static constexpr auto has_denorm_loss   = true;
```
Comment (CongMa13) [inline library/include/rocwmma/internal/float8.hpp:505]: Not quite sure about `has_denorm_loss`

File: test/gemm/gemm_predicates_base.hpp [hunk 3]
```diff
@@ -113,101 +113,141 @@ namespace rocwmma
...
+                  && (BlockK >= 8u)
                   && (BlockK % 8u == 0u)),
 
             // Older gfx908 arch has half BlockK on bfloat16_t
             // BlockM/N = 16; Block K >= 8
             // BlockM/N = 32; Block K >= 4
             Gfx908BF16BlockSizeTest
-            = !((bool)TestTraits::InputType::IsBFloat16 && (bool)TestTraits::Arch::IsGfx908)
-              || ((bool)TestTraits::BlockSizes::isBlockMN16 && (BlockK >= 8u)
+            = !((bool)TestTraits::InputType::IsBFloat16
```
Comment (CongMa13) [inline test/gemm/gemm_predicates_base.hpp]: Are these line breaks added manually?

They broke the clang-format code pattern. 

The git hook script will change them back if I commit this file later.

File: library/include/rocwmma/internal/float8.hpp [hunk 4]
```diff
@@ -24,948 +24,1047 @@
...
 
-// all + operator overloading with mixed types
-// mixed types, always converts to f32, does computation in f32, and returns float
-inline ROCWMMA_HOST_DEVICE float operator+(const float fa, rocwmma_f8 b)
+// Unary sign inversion
+ROCWMMA_FP8_FNUZ_VISIBILITY inline hip_fp8_e4m3_fnuz operator-(hip_fp8_e4m3_fnuz a)
 {
-    return (fa + float(b));
+    // Special case for 0 -> avoid flipping sign to NaN
+    return a.__x == __hip_fp8_storage_t{0} ? a : hip_fp8_e4m3_fnuz{a.__x ^ 0x80};
```
Comment (CongMa13) [inline library/include/rocwmma/internal/float8.hpp]: Should avoid converting -(nan) to 0 as well.

File: library/include/rocwmma/internal/float8.hpp [hunk 5]
```diff
@@ -24,948 +24,1047 @@
...
+            return make_hip_fp8_e4m3_from_bits(0x30);
+        }
+        static constexpr hip_fp8_e4m3 quiet_NaN()
+        {
+            return make_hip_fp8_e4m3_from_bits(0x7F);
+        }
+        static constexpr hip_fp8_e4m3 signaling_NaN()
+        {
+            return make_hip_fp8_e4m3_from_bits(0x7F);
+        }
```
Comment (CongMa13) [inline library/include/rocwmma/internal/float8.hpp:548]: Line 503: has_signaling_NaN = false;

e4m3 does not have a signaling nan

File: library/include/rocwmma/internal/float8.hpp [hunk 6]
```diff
@@ -24,948 +24,1047 @@
...
+        {
+            return make_hip_fp8_e5m2_from_bits(0x38);
+        }
+        static constexpr hip_fp8_e5m2 infinity()
+        {
+            return make_hip_fp8_e5m2_from_bits(0x7C);
+        }
+        static constexpr hip_fp8_e5m2 signaling_NaN()
+        {
+            return make_hip_fp8_e5m2_from_bits(0x7F);
```
Comment (CongMa13) [inline library/include/rocwmma/internal/float8.hpp:613]: Line 565: has_signaling_NaN = false;
e5m2 does not have a signaling nan.

File: test/gemm/gemm_predicates_base.hpp (hunk 3 above)
Comment (cgmillette) [reply test/gemm/gemm_predicates_base.hpp]: I'll run it again

**Review by CongMa13 — APPROVED** [review]
LGTM

**Pull request merged**